SRE White paper - openshift cluster (#12)
bcgov/platform-services-sre: docs/white-paper.md

# SITE RELIABILITY ENGINEERING (SRE) AT BRITISH COLUMBIA WITHIN THE PLATFORM AS A SERVICE (PAAS) – WHITE PAPER

This white paper is more of research analysis of British Columbia’s environment and identifying how to setup SRE related activities. This paper also provides some information about some common SRE principles and procedures. The paper will also provide a walk through of an example Product Team setting SRE objectives, indicators, monitoring, and observability with BC PaaS. 
## IDENTIFICATION OF ENVIRONMENTS FOR THE PAAS
It is important to notate that not all environments are created equal but if these cluster have core services or dependences then these need to be identified. Below is a list of environments currently identified for the BC PaaS.
•	Gold Cluster
•	Silver Cluster
•	Emerald Cluster
•	KLab Cluster

> •	KLab2 Cluster

Reviewer: ```suggestion
•	KLab2 Cluster
•	CLab Cluster
```

•	ROSA Cluster – **In Progress**

## GOALS AND VISION FOR EACH ENVIRONMENT
The goals and the vision for each environment is important. As this helps identify WHO IS THE CUSTOMER. There are two main categories for these environments:
•	Utilized by Product Teams
•	Utilized for Development and Test
**Note:** So, there should be more categories broken down between Gold Cluster, Silver Cluster, and Emerald, as the environments identify a different type of customer. 
For now, based on the information lets break apart the environments where the customer is the product team. 

**Product Team Environments:**
•	Gold Cluster
•	Silver Cluster
•	Emerald Cluster
•	ROSA Cluster – **In Progress**

**Platform Team Environments:**
•	KLab Cluster

> •	KLab2 Cluster

Reviewer: ```suggestion
•	KLab2 Cluster
•	CLab Cluster
```

So now that the environments are identified to the customer. Let’s identify the role in which each environment is used at a high level.
Product Team Environments – Utilized by Product Teams to host applications and services. 
Platform Teams Environments – Utilized for developing, testing, and validating services provided as part of the PaaS. 
So, with the identification of the environments lets identify an overall vision for the PaaS for the sake of the paper. 
“We want to provide a Platform to our customers that allow highly availability, scalability, and reliability across British Columbia meeting the mission needs”. 
IDENTIFY HIGH SERVICE LEVEL OBJECTIVES (SLO)S BY ENVIRONMENT
This section will discuss some of the high service level objectives that were identified by the PaaS between the DXE team and the Platform Team by environment. These are not set in stone as this is a living document that could change. 

**Current Silver Cluster SLOs**
•	90.0% availability for a single-node application deployment
•	99.5% availability for multi-node application deployments

**Current Gold Cluster SLOs**

> •	99.5% availability for multi-node application deployments

Reviewer: ```suggestion
•	99.95% availability for application deployments with a geographic failover
```

> **Current DXE Team SLO against the Cluster API Availability**

Reviewer 2: ```suggestion
**Current DXC Team SLO against the Cluster API Availability**
```

Utilization – Identify the percentage of the time that a Pod is being utilize on a Node within the cluster. 
Saturation – Identifying when services or the application is being access (aka queued). For example, if the service is an Application Program Interface (API) server such as Spring Boot and the Payload is taking a while to return because it is processing other requests. This can come in the form of utilization metrics. 
Errors – These are Errors that are actually “reported” against the service or application. 
Currently BC PaaS allows for the collection of the USE Method. Which can const of items such as:
•	Resource utilization of Memory and CPU. 
However, disk utilization I am currently unsure of as this is a on prem cluster and needs to be validated that the below query examples can be ran depending on permission issues from a client perspective. Below references some “Prometheus” queries that be used. Prometheus is one of the tool options that can be utilized. 
These are the standard metrics identified for Product Team. Golden Signals is a good place to start. These are metrics that can be identified no matter what the application is. Let’s look into Golden Signals. 
 

> The above is a high-level graphic of metrics to use for USE Method

Reviewer 2: What graphic?

Examples that can be utilize in Prometheus below for the Use Method.

## CPU Utilization:
```
1 - avg(rate(node_cpu{job="default/node-exporter",mode="idle"}[1m]))
```

OpenShift Example:
```
pod:container_cpu_usage:sum
```

CPU Saturation:
```
sum(node_load1{job="default/node-exporter"}) / sum(node:node_num_cpu:sum)
```

OpenShift Example:
```
sum by (pod,namespace) (rate(container_cpu_cfs_throttled_seconds_total{container=!""}[5m]))
```

Memory Utilization:
```
1 - sum( node_memory_MemFree{job=“…”} + node_memory_Cached{job=“…”} + node_memory_Buffers{job=“…”} ) / sum(node_memory_MemTotal{job=“…”})
```

OpenShift Example:
```
sum by(pod,namespace) (container_memory_working_set_bytes{container="",pod!=""})
```

Memory Saturation:
```
1e3 * sum( rate(node_vmstat_pgpgin{job=“…”}[1m]) + rate(node_vmstat_pgpgout{job=“…”}[1m])) )
```

OpenShift Example:
```

> max by (pod,namespace) (container_memory_working_set_bytes{container!=""} / on (pod,namespace,container)

Reviewer 2: This query doesn't look right?

We don't have swap on our clusters, so there should be no paging of memory. Once its full it gets OOM Killed.

Reviewer 2: pagein / pageout are not necessarily related to swap, should be fine to leave in. https://serverfault.com/a/270324/27378
